frontend/wasm/src/component/canon_abi_utils.rs

```diff
@@ -0,0 +1,175 @@
+use midenc_dialect_arith::ArithOpBuilder;
+use midenc_dialect_hir::HirOpBuilder;
+use midenc_hir::{AddressSpace, Builder, PointerType, SmallVec, SourceSpan, Type, ValueRef};
+
+use crate::{error::WasmResult, module::function_builder_ext::FunctionBuilderExt, WasmError};
+
+/// Recursively loads primitive values from memory based on the component-level type following the
+/// canonical ABI loading algorithm from
+/// https://github.com/WebAssembly/component-model/blob/main/design/mvp/CanonicalABI.md#loading
+pub fn load<B: ?Sized + Builder>(
+    fb: &mut FunctionBuilderExt<B>,
+    base_ptr: ValueRef,
+    ty: &Type,
+    offset: &mut u32,
+    values: &mut SmallVec<[ValueRef; 8]>,
+    span: SourceSpan,
+) -> WasmResult<()> {
+    // Align offset to the type's alignment requirement
+    let alignment = ty.min_alignment() as u32;
+    *offset = align_to(*offset, alignment);
+
+    match ty {
+        // Primitive types are loaded directly
+        Type::I1
+        | Type::I8
+        | Type::U8
+        | Type::I16
+        | Type::U16
+        | Type::I32
+        | Type::U32
+        | Type::I64
+        | Type::U64
+        | Type::Felt => {
+            let ptr_type =
+                Type::from(PointerType::new_with_address_space(ty.clone(), AddressSpace::Byte));
+
+            let src_offset = fb.i32(*offset as i32, span);
+            let src_addr = fb.add_unchecked(base_ptr, src_offset, span)?;
+            let src_ptr = fb.inttoptr(src_addr, ptr_type, span)?;
+            let value = fb.load(src_ptr, span)?;
+            values.push(value);
+
+            *offset += ty.size_in_bytes() as u32;
+        }
+
+        // Struct types are loaded field by field
+        Type::Struct(struct_ty) => {
+            // For each field in the struct, use the pre-calculated field offset
+            let struct_base_offset = *offset;
+            for field in struct_ty.fields() {
+                // Use the field's offset within the struct, plus the struct's base offset
+                let mut field_offset = struct_base_offset + field.offset;
+
+                // Recursively load the field
+                load(fb, base_ptr, &field.ty, &mut field_offset, values, span)?;
```

Comment:
Is there any reason why we're passing `base_ptr` + and offset here, rather than just aligning `base_ptr` for a given type, and then passing in the aligned pointer?

The reason I'm asking, is because otherwise we're going to compute incorrect addresses for the load, because alignments are going to be computed from the current offset, rather than the address of the load.

Ultimately, I don't see much value in passing an offset here; we also don't need to maintain one here AFAICT, because the field offsets for `struct_ty` are already computed to take into account the size of the fields + alignment.

Reply:
Good question! I think it works because I ignore the offset parameter when processing a struct. But yeah, I removed the offset altogether. Thank you!